toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -2,20 +2,65 @@
 // Exceptions. See /LICENSE for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-#include <utility>
+#include <cstdint>
 
 #include "toolchain/semantics/semantics_context.h"
+#include "toolchain/semantics/semantics_node.h"
+#include "toolchain/semantics/semantics_node_kind.h"
 
 namespace Carbon {
 
-auto SemanticsHandleIndexExpressionStart(SemanticsContext& context,
-                                         ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpressionStart");
+auto SemanticsHandleIndexExpressionStart(SemanticsContext& /*context*/,
+                                         ParseTree::Node /*parse_node*/)
+    -> bool {
+  // Leave the expression on the stack for IndexExpression.
+  return true;
 }
 
 auto SemanticsHandleIndexExpression(SemanticsContext& context,
                                     ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpression");
+  auto index_node_id = context.node_stack().PopExpression();
+  auto index_node = context.semantics_ir().GetNode(index_node_id);
+  auto name_node_id = context.node_stack().PopExpression();
+  auto name_node = context.semantics_ir().GetNode(name_node_id);
+  auto name_type_id =
+      context.semantics_ir().GetTypeAllowBuiltinTypes(name_node.type_id());
+  auto name_type_node = context.semantics_ir().GetNode(name_type_id);
+
+  if (name_type_node.kind() == SemanticsNodeKind::TupleType &&
+      index_node.kind() == SemanticsNodeKind::IntegerLiteral) {
+    auto index_val = *(context.semantics_ir()
+                           .GetIntegerLiteral(index_node.GetAsIntegerLiteral())
+                           .getRawData());
```

Comment:
`getRawData` returns a pointer to the start of the `APInt` storage, which can in general be an array of multiple `uint64_t`s. This is now only looking at the first word of the `APInt` data, which means the comparison below won't do the right thing for (eg) `0x10000000000000000`.

Instead, you can compare the `APInt` value itself:
```suggestion
    const auto& index_val = *(context.semantics_ir()
                                  .GetIntegerLiteral(index_node.GetAsIntegerLiteral()));
```

(Additional suggestions below to adapt the code to work with `APInt`.)